### PR TITLE
feat(state network): create StateContentValue type

### PR DESCRIPTION
**Pull request by morph-dev**
### What was wrong?

We didn't have any content-value types for the state network.

### How was it fixed?

Created minimal version of all required types. More functionality will be added as they are needed.

### To-Do

- [x] Add tests
- [ ] Clean up commit history and use [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).

### CODE REVIEW

**Comment by morph-dev** [comment]
I still have to create test vectors and add tests, but would like some initial feedback.

File: ethportal-api/src/types/state_trie/witness.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+use ssz_derive::{Decode, Encode};
+use ssz_types::{typenum, VariableList};
+
+use super::nibbles::Nibbles;
+
+/// The RLP encoding of a trie node.
+pub type EncodedTrieNode = VariableList<u8, typenum::U1024>;
+
+/// The ordered list of trie nodes. Together the make the path in a trie, first node being the
```
Comment (ogenev) [inline ethportal-api/src/types/state_trie/witness.rs:9]:
```suggestion
/// The ordered list of trie nodes. Together they make the path in a trie, first node being the
```

**Review by ogenev — APPROVED** [review]
Looks solid to me 🚀 .

File: ethportal-api/src/types/content_value/state.rs [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+pub struct ContractStorageTrieNodeWithProof {
+    /// A contract storage trie node with a proof.
+    pub proof: ContractStorageTrieWitness,
+    /// A block at which the proof is anchored.
+    pub block_hash: H256,
+}
+
+pub type ByteCode = VariableList<u8, typenum::U65536>;
+
+/// A content value type, used when retriving contract's bytecode.
```
Comment (njgheorghita) [inline ethportal-api/src/types/content_value/state.rs:97]: `retrieving`

File: ethportal-api/src/types/content_value/state.rs [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+/// A content value type, used when offering a trie node from the contract storage trie.
+#[derive(Clone, Debug, Decode, Encode, Eq, PartialEq)]
+pub struct ContractStorageTrieNodeWithProof {
+    /// A contract storage trie node with a proof.
+    pub proof: ContractStorageTrieWitness,
+    /// A block at which the proof is anchored.
+    pub block_hash: H256,
+}
+
+pub type ByteCode = VariableList<u8, typenum::U65536>;
```
Comment (njgheorghita) [inline ethportal-api/src/types/content_value/state.rs:95]: maybe just a small comment explaining why this max size was chosen

File: ethportal-api/src/types/content_value/state.rs [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+    use rstest::rstest;
+    use serde_json::Value;
+
+    use crate::{
+        types::state_trie::{nibbles::Nibbles, witness::TrieWitness},
+        utils::bytes::hex_decode,
+    };
+
+    use super::*;
+
```
Comment (njgheorghita) [inline ethportal-api/src/types/content_value/state.rs:128]: This seems like a good place to explain where the test values were sourced from - usually a github link (that contains some commit hash) is ideal

**Review by njgheorghita — APPROVED** [review]
Lookin' good!

File: ethportal-api/src/types/content_value/state.rs (hunk 4 above)
Comment (morph-dev) [reply ethportal-api/src/types/content_value/state.rs:123]: I just updated tests with new test vectors that are being discussed in new [PR](https://github.com/ethereum/portal-network-specs/pull/271). I will wait with merging this PR until the other one is merged and then I will add a link here.

**Review by pipermerriam — COMMENTED** [review]
No need to block this because of this comment, but it's something probably worth doing soon.

Our test vectors need to be importable via code.  Preferably something like embedding a git submodule at a specific commit and reading the files from that submodule.  Without this, these test vectors will inevitably become outdated because nobody copied in the new values...

And I suspect that may require some basic work standardizing how the test vectors are written and stored.

**Comment by KolbyML** [comment]
Standardized tests go in this https://github.com/ethereum/portal-spec-tests repo

**Review by KolbyML — APPROVED** [review]
:shipit:

**Comment by morph-dev** [comment]
I converted test vectors to .yaml files and added them to the https://github.com/ethereum/portal-spec-tests repo.

However, I'm not sure how I can access them from this repo? I know how to add repo as a dependency, but not how to read files from there. Any advice?

I'm fine with merging this as is, and followup later

**Comment by pipermerriam** [comment]
> However, I'm not sure how I can access them from this repo? I know how to add repo as a dependency, but not how to read files from there. Any advice?

I don't know how the team will want to do this, but in my past lives as a developer, we added the tests repo as a git submodule, which requires that CI be setup to pull down the submodules so that the files are available during test runs.

**Comment by KolbyML** [comment]
I looked into it a little bit and I think the best options are
- what piper suggested using a git submoduals
- or using a cargo make task https://github.com/sagiegurari/cargo-make?tab=readme-ov-file#tasks-dependencies-and-aliases which we could make when cargo test is run that portal-spec-tests is pulled

I haven't looked too much into git submoduals but ideally these tests are only pulled in if someone is using them aka running tests. So maybe this looks like a combination of both of the things I listed or just one of them.

I am :+1: for handling this in a follow up PR

**Comment by morph-dev** [comment]
I believe that git submodules are the best way forward, but I would definitely leave that work for a separate PR.

**Comment by ogenev** [comment]
Going to merge this, let's iterate on it in future PRs.

**Pull request merged**